library/network/src/lib/y2firewall/firewalld.rb

```diff
@@ -176,6 +176,8 @@ def write_only
       read unless read?
       apply_zones_changes!
       apply_attributes_changes!
+      # remove zones that is not in new zones map
+      (api.zones - zones.map(&:name)).each { |n| api.delete_zone(n) }
```

Comment:
**Note:** api calls are slow, and we avoid them as much as possible.  You can use current_zone_names already

Also, this responsibility was originally added to the `apply_zones_changes!` method hidden it from the object itself.
```ruby
      zones.each do |zone|
        api.create_zone(zone.name) unless current_zone_names.include?(zone.name)
        zone.apply_changes! if zone.modified?
      end
      current_zone_names.each do |name|
        api.delete_zone(name) if !zones.any? { |z| z.name == name }
      end
```

```